src/main/java/javaslang/control/Validation.java

```diff
@@ -0,0 +1,276 @@
+    }
+
+    default <U> Validation<U,T> leftMap(Function<? super E,? extends U> f) {
+        if(isInvalid()) {
+            E v = this.error();
+            return Validation.invalid(f.apply(v));
+        } else {
+            return (Valid<U,T>) this;
+        }
+    }
+
+    default <U> Validation<List<E>,U> ap(Validation<List<E>, ? extends Function<? super T,? extends U>> v) {
+        if(isValid() && v.isValid()) {
+            return valid(() -> {
+                Function<? super T,? extends U> f = ((Valid<E,? extends Function<? super T,? extends U>>) v).get();
+                return f.apply(this.get());
+            });
+        } else if(isValid() && v.isInvalid()) {
+            List<E> errors = v.error();
+            return invalid(errors);
+        } else if(isInvalid() && v.isValid()) {
+            E error = this.error();
+            return invalid(List.of(error));
+        } else {
+            return invalid(() -> {
+                List<E> errors = v.error();
+                E e = this.error();
+                return errors.append(e);
+            });
+        }
+    }
+
+    @Override
+    default <U> Validation<List<E>,U> ap(Kind<Validation<?, ?>, List<E>, ? extends Function<? super T, ? extends U>> f) {
+        return ap((Validation<List<E>,? extends Function<? super T,? extends U>>) ((Object) f));
+    }
+
+    default <U> ValidationBuilder<E,T,U> combine(Validation<E,U> v) {
+        return new ValidationBuilder<>(this, v);
+    }
+
+
+    final class Valid<E,T> implements Validation<E,T> {
+
+        private final T value;
+
+        /**
+         * Construct a {@code Valid}
+         * @param value The value of this success
+         */
+        public Valid(T value) {
+            this.value = value;
+        }
+
+        @Override
+        public boolean isValid() {
+            return true;
+        }
+
+        @Override
+        public boolean isInvalid() {
+            return false;
+        }
+
+        @Override
+        public void forEach(Consumer<? super T> f) {
+            f.accept(get());
+        }
+
+        @Override
+        public T get() {
+            return value;
+        }
+
+        @Override
+        public E error() throws RuntimeException {
+            throw new NoSuchElementException("error of 'valid' Validation");
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            return (obj == this) || (obj instanceof Valid && Objects.equals(value, ((Valid<?,?>) obj).value));
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hashCode(value);
+        }
+
+        @Override
+        public String toString() {
+            return "Valid(" + value + ")";
+        }
+
+    }
+
+    final class Invalid<E,T> implements Validation<E,T> {
+
+        private final E error;
+
+        /**
+         * Construct an {@code Invalid}
+         * @param error The value of this error
+         */
+        public Invalid(E error) {
```

Comment:
Please make the constructor `private` to be consistent with the other controls. Instances are created via the Validation API.